crates/ark/src/lsp/code_action.rs

```diff
@@ -0,0 +1,112 @@
+//! Code actions
+//!
+//! These are contextual light bulbs that appear when the user's cursor is at a particular
+//! position. They allow for small context specific quick fixes, refactors, documentation
+//! generation, and other small code adjustments.
+//!
+//! Modeled after rust-analyzer's blog post:
+//! https://rust-analyzer.github.io/blog/2020/09/28/how-to-make-a-light-bulb.html
+
+use std::collections::HashMap;
+
+use tower_lsp::lsp_types::CodeAction;
+use tower_lsp::lsp_types::CodeActionKind;
+use tower_lsp::lsp_types::CodeActionOrCommand;
+use tower_lsp::lsp_types::CodeActionResponse;
+use tower_lsp::lsp_types::DocumentChanges;
+use tower_lsp::lsp_types::OneOf;
+use tower_lsp::lsp_types::OptionalVersionedTextDocumentIdentifier;
+use tower_lsp::lsp_types::TextDocumentEdit;
+use tower_lsp::lsp_types::TextEdit;
+use tower_lsp::lsp_types::WorkspaceEdit;
+use tree_sitter::Range;
+use url::Url;
+
+use crate::lsp::capabilities::Capabilities;
+use crate::lsp::code_action::roxygen::roxygen_documentation;
+use crate::lsp::documents::Document;
+
+mod roxygen;
+
+/// A small wrapper around [CodeActionResponse] that make a few things more ergonomic
+pub(crate) struct CodeActions {
+    response: CodeActionResponse,
+}
+
+pub(crate) fn code_actions(
+    uri: &Url,
+    document: &Document,
+    range: Range,
+    capabilities: &Capabilities,
+) -> CodeActionResponse {
+    let mut actions = CodeActions::new();
+
+    roxygen_documentation(&mut actions, uri, document, range, capabilities);
+
+    actions.into_response()
+}
+
+pub(crate) fn code_action(title: String, kind: CodeActionKind, edit: WorkspaceEdit) -> CodeAction {
+    CodeAction {
+        title,
+        kind: Some(kind),
+        edit: Some(edit),
+        diagnostics: None,
+        command: None,
+        is_preferred: None,
+        disabled: None,
+        data: None,
+    }
+}
+
+/// Creates a common kind of `WorkspaceEdit` composed of one or more `TextEdit`s to
+/// apply to a single document
+pub(crate) fn code_action_workspace_text_edit(
+    uri: Url,
```

Comment:
There is a lot of boilerplate to creating an LSP `CodeAction`, so I've wrapped up what I hope to be a common case in a few helpers. Time will tell, and we can adjust as required later